### PR TITLE
Spill Statistics

### PR DESCRIPTION
**Pull request by madsbk**
Implements statistics of spilling, enable by setting `CUDF_SPILL_STATS=2`.

It is possible to access the statistics through the spill manager like:
```python
>>> get_global_manager().statistics
<SpillStatistics level=1>
```

To have each work in dask print spill statistics, do something like:
```python
    def spill_info():
        from cudf.core.buffer.spill_manager import get_global_manager
        manager = get_global_manager()
        stat = manager.statistics
        ret = f"spill_info() - get_global_manager: {manager}\n"
        ret += f"{stat}"
        print(ret)

    client.submit(spill_info)
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by wence- — CHANGES_REQUESTED** [review]
Some minor things, but overall I think this looks good.

File: python/cudf/cudf/core/buffer/spill_manager.py [hunk 1]
```diff
@@ -46,6 +49,144 @@ def get_rmm_memory_resource_stack(
...
 
 
+class SpillStatistics:
+    """Gather spill statistics
+
+    Levels of information gathered:
+      0  - disabled (no overhead).
+      1+ - duration and number of bytes spilled (very low overhead).
+      2+ - a traceback for each time a spillable buffer is exposed
+           permanently (potential high overhead).
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spill_manager.py:59]: Yes, I think that could be useful.

File: python/cudf/cudf/core/buffer/spill_manager.py [hunk 2]
```diff
@@ -46,6 +49,144 @@ def get_rmm_memory_resource_stack(
...
+        """Log an expose event
+
+        We track logged exposes by grouping them by their traceback such
+        that `self.exposes` maps tracebacks (as strings) to their logged
+        data (as `Expose`).
+
+        Parameters
+        ----------
+        buf : spillabe-buffer
+            The buffer being exposed.
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spill_manager.py:135]: I agree, but this is how it is done throughout cudf. Of cause, we could change this? `numpydoc` supports [non-typed parameters](https://numpydoc.readthedocs.io/en/latest/format.html#parameters).

cc. @shwina @galipremsagar

File: python/cudf/cudf/core/buffer/spill_manager.py [hunk 3]
```diff
@@ -46,6 +49,144 @@ def get_rmm_memory_resource_stack(
...
+            ret += textwrap.indent(
+                (
+                    f"exposed {s.count} times, "
+                    f"total: {format_bytes(s.total_nbytes)}, "
+                    f"spilled: {format_bytes(s.spilled_nbytes)}, "
+                    f"traceback:\n{s.traceback}"
+                ),
+                prefix=" " * 4,
+            )
+        return ret[:-1]  # Remove last `\n`
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spill_manager.py:187]: True, but I like the convenience of `textwrap.indent` and I would also have to strip the newlines in `s.traceback`.

File: python/cudf/cudf/core/buffer/spill_manager.py [hunk 4]
```diff
@@ -46,6 +49,144 @@ def get_rmm_memory_resource_stack(
...
+        count: int = 1
+        total_nbytes: int = 0
+        spilled_nbytes: int = 0
+
+    spills_totals: Dict[Tuple[str, str], Tuple[int, float]]
+
+    def __init__(self, level) -> None:
+        self.lock = threading.Lock()
+        self.level = level
+        self.spills_totals = defaultdict(lambda: (0, 0))
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spill_manager.py:98]: Done

[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spill_manager.py:98]: Done

File: python/cudf/cudf/core/buffer/spill_manager.py [hunk 5]
```diff
@@ -46,6 +49,144 @@ def get_rmm_memory_resource_stack(
...
+        if self.level == 0:
+            return ret[:-1] + " N/A"
+
+        # Print spilling stats
+        ret += "  Spilling (level >= 1):"
+        if len(self.spills_totals) == 0:
+            ret += " None"
+        ret += "\n"
+        for (src, dst), (nbytes, time) in self.spills_totals.items():
+            ret += f"    {src} => {dst}: {format_bytes(nbytes)} in {time}s\n"
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/core/buffer/spill_manager.py:168]: Done

File: python/cudf/cudf/tests/test_spilling.py [hunk 6]
```diff
@@ -475,3 +489,76 @@ def test_memoryview_slice(manager: SpillManager, dtype):
...
+    assert stat.count == 1
+    assert stat.total_nbytes == buffers[0].nbytes
+    assert stat.spilled_nbytes == 0
+
+    # Expose all 10 buffers
+    for i in range(10):
+        buffers[i].ptr
+
+    # The rest should accumulate to a single stat
+    assert len(manager.statistics.exposes) == 2
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/tests/test_spilling.py:543]: Added comment

File: python/cudf/cudf/tests/test_spilling.py [hunk 7]
```diff
@@ -475,3 +489,76 @@ def test_memoryview_slice(manager: SpillManager, dtype):
...
+    buffers[0].ptr
+    assert len(manager.statistics.exposes) == 1
+    stat = list(manager.statistics.exposes.values())[0]
+    assert stat.count == 1
+    assert stat.total_nbytes == buffers[0].nbytes
+    assert stat.spilled_nbytes == 0
+
+    # Expose all 10 buffers
+    for i in range(10):
+        buffers[i].ptr
```
[the remark this answers is not in the document]
Comment (madsbk) [reply python/cudf/cudf/tests/test_spilling.py:540]: It shouldn't have any consequences if it goes over the limit. `spill_device_limit` is a _soft_ limit, which I have added to the docs: d0271a51f85a7d71ad3a0982d8598f662d107343

**Comment by madsbk** [comment]
@wence-, thanks for the review! I think it is ready for another round.

File: python/cudf/cudf/core/buffer/spill_manager.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wence-) [reply python/cudf/cudf/core/buffer/spill_manager.py]: Fair.

**Review by wence- — APPROVED** [review]
Very minor documentation quibbles, otherwise thanks!

**Comment by madsbk** [comment]
rerun tests

**Comment by madsbk** [comment]
@gpucibot merge

**Comment by madsbk** [comment]
> Very minor documentation quibbles, otherwise thanks!

Thanks @wence- !

**Pull request merged**